floodsub.go

```diff
@@ -2,454 +2,46 @@ package floodsub
-		if !p.subscribedToMsg(pmsg) {
-			log.Warning("received message we didn't subscribe to. Dropping.")
-			continue
-		}
-
-		msg := &Message{pmsg}
-		vals := p.getValidators(msg)
-		p.pushMsg(vals, rpc.from, msg)
-	}
-}
-
-// msgID returns a unique ID of the passed Message
-func msgID(pmsg *pb.Message) string {
-	return string(pmsg.GetFrom()) + string(pmsg.GetSeqno())
-}
-
-// pushMsg pushes a message performing validation as necessary
-func (p *PubSub) pushMsg(vals []*topicVal, src peer.ID, msg *Message) {
-	if len(vals) > 0 {
-		// validation is asynchronous and globally throttled with the throttleValidate semaphore.
-		// the purpose of the global throttle is to bound the goncurrency possible from incoming
-		// network traffic; each validator also has an individual throttle to preclude
-		// slow (or faulty) validators from starving other topics; see validate below.
-		select {
-		case p.validateThrottle <- struct{}{}:
-			go func() {
-				p.validate(vals, src, msg)
-				<-p.validateThrottle
-			}()
-		default:
-			log.Warningf("message validation throttled; dropping message from %s", src)
-		}
-		return
-	}
-
-	p.maybePublishMessage(src, msg.Message)
+func (fs *FloodSubRouter) Protocols() []protocol.ID {
+	return []protocol.ID{FloodSubID}
 }
 
-// validate performs validation and only sends the message if all validators succeed
-func (p *PubSub) validate(vals []*topicVal, src peer.ID, msg *Message) {
-	ctx, cancel := context.WithCancel(p.ctx)
-	defer cancel()
-
-	rch := make(chan bool, len(vals))
-	rcount := 0
-	throttle := false
-
-loop:
-	for _, val := range vals {
-		rcount++
-
-		select {
-		case val.validateThrottle <- struct{}{}:
-			go func(val *topicVal) {
-				rch <- val.validateMsg(ctx, msg)
-				<-val.validateThrottle
-			}(val)
-
-		default:
-			log.Debugf("validation throttled for topic %s", val.topic)
-			throttle = true
-			break loop
-		}
-	}
-
-	if throttle {
-		log.Warningf("message validation throttled; dropping message from %s", src)
-		return
-	}
-
-	for i := 0; i < rcount; i++ {
-		valid := <-rch
-		if !valid {
-			log.Warningf("message validation failed; dropping message from %s", src)
-			return
-		}
-	}
-
-	// all validators were successful, send the message
-	p.sendMsg <- &sendReq{
-		from: src,
-		msg:  msg,
-	}
+func (fs *FloodSubRouter) Attach(p *PubSub) {
+	fs.p = p
 }
 
-func (p *PubSub) maybePublishMessage(from peer.ID, pmsg *pb.Message) {
-	id := msgID(pmsg)
-	if p.seenMessage(id) {
-		return
-	}
+func (fs *FloodSubRouter) AddPeer(peer.ID, protocol.ID) {}
 
-	p.markSeen(id)
+func (fs *FloodSubRouter) RemovePeer(peer.ID) {}
 
-	p.notifySubs(pmsg)
+func (fs *FloodSubRouter) HandleRPC(rpc *RPC) {}
 
-	p.publishMessage(from, pmsg)
-}
-
-func (p *PubSub) publishMessage(from peer.ID, msg *pb.Message) {
+func (fs *FloodSubRouter) Publish(from peer.ID, msg *pb.Message) {
 	tosend := make(map[peer.ID]struct{})
 	for _, topic := range msg.GetTopicIDs() {
-		tmap, ok := p.topics[topic]
+		tmap, ok := fs.p.topics[topic]
```

Comment:
Yes, but that's a bit further down the road. 
This is step 1, so that we can write the first router or two without impedance mismatch.
And once we have a clear idea of what kind of state we want to access (and how), we can make an api for it and restrict it.